### PR TITLE
[KYUUBI #3444] Support the planOnly mode of kyuubi spark engine support SQL lineage

### LINKED ISSUE #3444: [FEATURE] The planOnly mode of kyuubi spark engine support SQL lineage output
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the feature

 Executing sql with the `PlanOnly` mode will get the output of SQL's lineage

### Motivation

_No response_

### Describe the solution

_No response_

### Additional context

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by iodone**
Closes #3444 

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/PlanOnlyStatement.scala [hunk 1]
```diff
@@ -136,10 +142,25 @@ class PlanOnlyStatement(
...
       case UnknownMode => throw unknownModeError(mode)
       case _ =>
         throw KyuubiSQLException(s"The operation mode $mode" +
           " doesn't support in Spark SQL engine.")
     }
   }
+
+  private def parseLineage(spark: SparkSession, plan: LogicalPlan): String = {
+    val analyzed = spark.sessionState.analyzer.execute(plan)
+    spark.sessionState.analyzer.checkAnalysis(analyzed)
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/PlanOnlyStatement.scala:157]: `optimized` has done this check ?

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/PlanOnlyStatement.scala [hunk 2]
```diff
@@ -136,10 +142,25 @@ class PlanOnlyStatement(
...
+    val optimized = spark.sessionState.optimizer.execute(analyzed)
+    val lineageParser = new LineageParser { def sparkSession = spark }
+    val lineage = lineageParser.parse(optimized)
+
+    val mapper = new ObjectMapper().registerModule(DefaultScalaModule)
+    val r = mapper.writeValueAsString(lineage)
+    r
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/PlanOnlyStatement.scala:164]: unnecessary variable

**Review by ulysses-you — APPROVED** [review]
looks a good feature, cc @yaooqinn @cfmcgrady

File: externals/kyuubi-spark-sql-engine/pom.xml [hunk 3]
```diff
@@ -51,6 +51,12 @@
             <version>${project.version}</version>
         </dependency>
 
+        <dependency>
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/pom.xml:54]: any idea about removing this from the engine？ It makes engine module hard for long-term maintenance and cross version support

Comment (iodone) [reply externals/kyuubi-spark-sql-engine/pom.xml:54]: @yaooqinn @ulysses-you @cfmcgrady 
https://github.com/apache/incubator-kyuubi/pull/3537#issuecomment-1260305095 Here just discussed the dependency with the plugin. @wForget suggested extracting the common module to reuse `SQLLineageParserHelper`, can you give some other advice?

Comment (ulysses-you) [reply externals/kyuubi-spark-sql-engine/pom.xml:54]: the lineage module does not shade the spark, so I think the worst case is the lineage can not work and the engine won't be broken ?

we can add try-catch for those code in engine and narrow the exception

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/PlanOnlyStatement.scala (hunk 1 above)
Comment (iodone) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/operation/PlanOnlyStatement.scala:157]: Call the`executeAndCheck` method in the `analyzed` stage when query execution, this method includes `checkAnalysis`, but here just call the `spark.sessionState.analyzer.execute(plan)`, so `optimized` don't do this check

File: externals/kyuubi-spark-sql-engine/pom.xml (hunk 3 above)
Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/pom.xml:54]: A spark engine plugin can depend on engine lib(not suggested, but ok) and spark libs for the compilation phase, but the engine lib can not do the opposite.

Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/pom.xml:54]: how about have a rule or logic inside the lineage plugin, the engine side can push down a config or something, then you determine the output

**Comment by ulysses-you** [comment]
the github action seems get sick.. every commit need approve to run : (

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/KyuubiSparkUtil.scala [hunk 4]
```diff
@@ -93,4 +96,13 @@ object KyuubiSparkUtil extends Logging {
     val runtimeVersion = SemanticVersion(runtimeSparkVer)
     (runtimeVersion.majorVersion, runtimeVersion.minorVersion)
   }
+
+  def parsePlan(spark: SparkSession, plan: LogicalPlan): String = {
+    import org.apache.kyuubi.plugin.lineage.helper.LineageParser
+    val lineageParser =
+      new LineageParser { def sparkSession = spark }
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/KyuubiSparkUtil.scala:103]: can we use reflect here ? then we can remove maven dependency

Comment (iodone) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/KyuubiSparkUtil.scala]: Using reflection is similar to setting the dependency of spark-lineage-plugin to `optional` and requires that this dependency logic be put into a separate file to pass all the unit tests.

File: externals/kyuubi-spark-sql-engine/pom.xml (hunk 3 above)
Comment (iodone) [reply externals/kyuubi-spark-sql-engine/pom.xml:54]: @yaooqinn Can we use `optional` dependencies here? Refer to the definition of maven `optional`: 

> The idea is that some of the dependencies are only used for certain features in the project and will not be needed if that feature isn't used

Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/pom.xml:54]: Can we just make the full functionality on the plugin side? IMHO, it is not necessary to include a dependency based on my experience to create more than one plugins

Comment (iodone) [reply externals/kyuubi-spark-sql-engine/pom.xml]: @yaooqinn I have removed the dependency with your advice

**Comment by wForget** [comment]
We can define a PlanOnlyStatement mode interface in Engine, and then implement it in the lineage plugin.

**Comment by iodone** [comment]
@yaooqinn @wForget This feature has been overlooked. A new version has recently been released. The engine has removed the strong dependency on the plugin by reflecting on the provider class of the plugin. Can you help me review it again?

**Comment by pan3793** [comment]
cc @ulysses-you

**Review by ulysses-you — APPROVED** [review]

**Comment by ulysses-you** [comment]
thanks, merging to master

**Pull request closed without merging**